### ISSUE TITLE
Expression ends prematurely in cider-last-sexp

### ISSUE THREAD

**Issue by arichiardi** [clojure-emacs/cider#1939]
## Expected behavior

It should return `nil` maybe?

## Actual behavior
It throws `forward-sexp: Scan error: "Containing expression ends prematurely", 1331, 1331`

## Steps to reproduce the problem

In the following expression and point as indicated, execute `cider-last-sexp`:

`(|System/getProperty "maven.local-repo")`

### CIDER version information
```
;; CIDER 0.15.0snapshot (latest master), nREPL 0.2.12
;; Clojure 1.9.0-alpha14, Java 1.8.0_x
```

### Emacs version

`GNU Emacs 24.5.1 (x86_64-pc-linux-gnu, GTK+ Version 3.18.9) of 2016-04-17 on lgw01-04, modified by Debian`

**Comment by bbatsov** [comment]
The actual problem seems to be in `clojure-backward-logical-sexp`.

**Comment by bbatsov** [comment]
On further inspection it seems that `backward-sexp` blows up every time it's called from the beginning of some form.  Not sure how best to approach this. @Malabarba any thoughts?

**Comment by bbatsov** [comment]
My ideas - either wrap all `backward-sexp` calls in `ignore-errors` or add some check whether we're at the beginning of a form. Probably the second idea is the better one, but as I said - I'm not quite certain about this.

**Comment by martinklepsch** [comment]
I'm also noticing a similar issue. It doesn't happen all the time but at some point my Emacs gets into a state from which point it happens whenever I run `M-x integrant-reset`, defined as:

```
(defun integrant-reset ()
  (interactive)
  (cider-interactive-eval "(integrant.repl/reset)"))
```
edebug has this to say:
```
Debugger entered--Lisp error: (scan-error "Containing expression ends prematurely" 24251 24251)
  apply(debug error (scan-error "Containing expression ends prematurely" 24251 24251))
  edebug(error (scan-error "Containing expression ends prematurely" 24251 24251))
  scan-sexps(24252 -1)
  forward-sexp(-1)
  backward-sexp(1)
  clojure-backward-logical-sexp(1)
  clojure-forward-logical-sexp(-1)
  cider--make-fringe-overlay(nil)
  #f(compiled-function (buffer value) #<bytecode 0x4078d131>)(#<buffer sqlite_impl.clj> ":resumed")
  #f(compiled-function (response) #<bytecode 0x4078d1c5>)((dict "id" "159" "ns" "cljdoc.storage.sqlite-impl" "session" "7edadf13-04ee-4ff3-b525-115340147663" "value" ":resumed"))
  #f(compiled-function (response) #<bytecode 0x4078d261>)((dict "id" "159" "ns" "cljdoc.storage.sqlite-impl" "session" "7edadf13-04ee-4ff3-b525-115340147663" "value" ":resumed"))
  nrepl--dispatch-response((dict "id" "159" "ns" "cljdoc.storage.sqlite-impl" "session" "7edadf13-04ee-4ff3-b525-115340147663" "value" ":resumed"))
  nrepl-client-filter(#<process nrepl-connection> "d2:id3:1593:out138:INFO [2019-03-26 10:56:35,347] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - cljdoc.server.release-monitor Stopping ReleaseMonitor\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out200:INFO [2019-03-26 10:56:35,353] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - org.eclipse.jetty.server.AbstractConnector Stopped ServerConnector@712b757e{HTTP/1.1,[http/1.1, h2c]}{0.0.0.0:8000}\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out199:INFO [2019-03-26 10:56:35,353] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - org.eclipse.jetty.server.handler.ContextHandler Stopped o.e.j.s.ServletContextHandler@7a8dc7a0{/,null,UNAVAILABLE}\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out14::reloading ()\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed18:changed-namespacesde2:id3:1589:repl-type3:clj7:session36:7edadf13-04ee-4ff3-b525-1153401476636:statusl5:stateeed2:id3:1593:out143:INFO [2019-03-26 10:56:35,382] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - cljdoc.server.system Starting :cljdoc/analysis-service nil\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out136:INFO [2019-03-26 10:56:35,389] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - cljdoc.server.system Starting :cljdoc/build-tracker\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out130:INFO [2019-03-26 10:56:35,390] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - cljdoc.server.system Starting :cljdoc/storage\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out138:INFO [2019-03-26 10:56:35,392] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - cljdoc.server.pedestal Starting pedestal on port 8000\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out138:INFO [2019-03-26 10:56:35,408] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - org.eclipse.jetty.server.Server jetty-9.4.0.v20161208\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out197:INFO [2019-03-26 10:56:35,418] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - org.eclipse.jetty.server.handler.ContextHandler Started o.e.j.s.ServletContextHandler@5d1b637f{/,null,AVAILABLE}\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out200:INFO [2019-03-26 10:56:35,419] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - org.eclipse.jetty.server.AbstractConnector Started ServerConnector@6e66736f{HTTP/1.1,[http/1.1, h2c]}{0.0.0.0:8000}\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out135:INFO [2019-03-26 10:56:35,420] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - org.eclipse.jetty.server.Server Started @4623834ms\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1593:out153:INFO [2019-03-26 10:56:35,421] nRepl-session-7edadf13-04ee-4ff3-b525-115340147663 - cljdoc.server.release-monitor Starting ReleaseMonitor (dry-run mode)\n7:session36:7edadf13-04ee-4ff3-b525-115340147663ed2:id3:1592:ns26:cljdoc.storage.sqlite-impl7:session36:7edadf13-04ee-4ff3-b525-1153401476635:value8::resumeded2:id3:1597:session36:7edadf13-04ee-4ff3-b525-1153401476636:statusl4:doneee")
  recursive-edit()
  edebug--recursive-edit(before)
  edebug--display-1(nil 0 before)
  edebug--display(nil 0 before)
  edebug-debugger(0 before nil)
  edebug-before(0)
  (edebug-after (edebug-before 0) 1 (integrant-reset))
  (lambda nil (edebug-after (edebug-before 0) 1 (integrant-reset)))()
  edebug-enter(edebug-anon0 nil (lambda nil (edebug-after (edebug-before 0) 1 (integrant-reset))))
  edebug-enter(edebug-anon0 nil (lambda nil (edebug-after (edebug-before 0) 1 (integrant-reset))))
  eval((edebug-enter 'edebug-anon0 nil (function (lambda nil (edebug-after (edebug-before 0) 1 (integrant-reset))))) nil)
  eval-expression((edebug-enter 'edebug-anon0 nil (function (lambda nil (edebug-after (edebug-before 0) 1 (integrant-reset))))))
  edebug-defun()
  funcall-interactively(edebug-defun)
  call-interactively(edebug-defun record nil)
  command-execute(edebug-defun record)
  helm-M-x(nil #("edebug-defun" 0 12 (match-part "edebug-defun")))
  funcall-interactively(helm-M-x nil #("edebug-defun" 0 12 (match-part "edebug-defun")))
  call-interactively(helm-M-x nil nil)
  command-execute(helm-M-x)
```
